Core/src/TrackFitting/GainMatrixUpdater.cpp

```diff
@@ -16,71 +19,70 @@ std::tuple<double, std::error_code> GainMatrixUpdater::visitMeasurement(
+
+    const auto K = (trackState.predictedCovariance * H.transpose() *
+                    (H * trackState.predictedCovariance * H.transpose() +
+                     calibratedCovariance)
+                        .inverse())
+                       .eval();
+
+    ACTS_VERBOSE("Gain Matrix K:\n" << K);
+
+    if (K.hasNaN()) {
+      error = (direction == NavigationDirection::Forward)
                   ? KalmanFitterError::ForwardUpdateFailed
                   : KalmanFitterError::BackwardUpdateFailed;  // set to error
-          return false;                                       // abort execution
-        }
-
-        trackState.filtered =
-            trackState.predicted + K * (calibrated - H * trackState.predicted);
-        trackState.filteredCovariance = (BoundSymMatrix::Identity() - K * H) *
-                                        trackState.predictedCovariance;
-        ACTS_VERBOSE(
-            "Filtered parameters: " << trackState.filtered.transpose());
-        ACTS_VERBOSE("Filtered covariance:\n" << trackState.filteredCovariance);
-
-        // calculate filtered residual
-        //
-        // FIXME: Without separate residual construction and assignment, we
-        //        currently take a +0.7GB build memory consumption hit in the
-        //        EventDataView unit tests. Revisit this once Measurement
-        //        overhead problems (Acts issue #350) are sorted out.
-        //
-        ParametersVector residual;
-        residual = calibrated - H * trackState.filtered;
-        ACTS_VERBOSE("Residual: " << residual.transpose());
-
-        chi2 = (residual.transpose() *
-                ((CovarianceMatrix::Identity() - H * K) * calibratedCovariance)
-                    .inverse() *
-                residual)
-                   .value();
-
-        ACTS_VERBOSE("Chi2: " << chi2);
-        return true;  // continue execution
-      });
+      return false;                                           // abort execution
+    }
+
+    trackState.filtered =
+        trackState.predicted + K * (calibrated - H * trackState.predicted);
+    trackState.filteredCovariance =
+        (BoundSymMatrix::Identity() - K * H) * trackState.predictedCovariance;
+    ACTS_VERBOSE("Filtered parameters: " << trackState.filtered.transpose());
+    ACTS_VERBOSE("Filtered covariance:\n" << trackState.filteredCovariance);
+
+    // calculate filtered residual
+    //
+    // FIXME: Without separate residual construction and assignment, we
+    //        currently take a +0.7GB build memory consumption hit in the
+    //        EventDataView unit tests. Revisit this once Measurement
+    //        overhead problems (Acts issue #350) are sorted out.
+    //
+    ParametersVector residual;
+    residual = calibrated - H * trackState.filtered;
+    ACTS_VERBOSE("Residual: " << residual.transpose());
+
+    chi2 = (residual.transpose() *
+            ((CovarianceMatrix::Identity() - H * K) * calibratedCovariance)
```

Comment:
A curiosity. Since now CovMatrix can be of different size why there is no size specifier here? Not sure how it cold actually work.

Reply:
H is the projector and has the dimension `eBoundSize x N`, where N is the measurement dimension. This should give an `NxN` matrix, which can be multiplied by `calibratedCovariance` which is also `NxN`.